src/strands/multiagent/swarm.py

```diff
@@ -0,0 +1,646 @@
+        }
+
+
+@dataclass
+class SwarmState:
+    """Current state of swarm execution."""
+
+    current_node: SwarmNode
+    task: str
+    completion_status: Status = Status.PENDING
+    shared_context: SharedContext = field(default_factory=SharedContext)
+    node_history: list[SwarmNode] = field(default_factory=list)
+    message_history: list[SwarmMessage] = field(default_factory=list)
+    iteration_count: int = 0
+    start_time: float = field(default_factory=time.time)
+    last_node_sequence: list[SwarmNode] = field(default_factory=list)
+    final_result: str | None = None
+    results: dict[str, NodeResult] = field(default_factory=dict)
+    accumulated_usage: Usage = field(default_factory=lambda: Usage(inputTokens=0, outputTokens=0, totalTokens=0))
+    accumulated_metrics: Metrics = field(default_factory=lambda: Metrics(latencyMs=0))
+    execution_count: int = 0
+    execution_time: int = 0
+
+    def should_continue(self, config: SwarmConfig) -> Tuple[bool, str]:
+        """Check if the swarm should continue.
+
+        Returns: (should_continue, reason)
+        """
+        elapsed = time.time() - self.start_time
+
+        # 1. Check completion status
+        if self.completion_status != Status.EXECUTING:
+            return False, f"completion_status_changed_to_{self.completion_status}"
+
+        # 2. Check handoff limit
+        if len(self.node_history) >= config.max_handoffs:
+            self.completion_status = Status.FAILED
+            return False, f"max_handoffs_reached_{config.max_handoffs}"
+
+        # 3. Check iteration limit
+        if self.iteration_count >= config.max_iterations:
+            self.completion_status = Status.FAILED
+            return False, f"max_iterations_reached_{config.max_iterations}"
+
+        # 4. Check timeout
+        if elapsed > config.execution_timeout:
+            self.completion_status = Status.FAILED
+            return False, f"execution_timeout_{config.execution_timeout}s"
+
+        # 5. Check for node ping-pong (nodes passing back and forth)
+        if len(self.last_node_sequence) >= config.ping_pong_check_nodes:
+            recent = self.last_node_sequence[-config.ping_pong_check_nodes :]
+            unique_nodes = len(set(recent))
+            if unique_nodes < config.ping_pong_min_unique_nodes:
+                self.completion_status = Status.FAILED
+                return (
+                    False,
+                    f"node_ping_pong_detected_{unique_nodes}_unique_in_{config.ping_pong_check_nodes}_recent",
+                )
+
+        return True, "continuing"
+
+    def increment_iteration(self, current_node: SwarmNode, config: SwarmConfig) -> None:
+        """Increment iteration count and track node usage."""
+        self.iteration_count += 1
+        self.last_node_sequence.append(current_node)
+
+        # Keep only the required number of nodes for ping-pong detection
+        if len(self.last_node_sequence) > config.ping_pong_check_nodes:
```

Comment:
Ah, ping-ponging I'm guessing is making sure that it's not revisiting the same nodes over and over